### ISSUE TITLE
wierd conversation with zephyr-7b-dpo-lora

### ISSUE THREAD

**Issue by njupopsicle** [huggingface/alignment-handbook#78]
![VX0N3YSQ@~VU@RPDR5@$KWE](https://github.com/huggingface/alignment-handbook/assets/84949179/cd22f16c-e160-43a9-b2cc-aec5b87b1ada)
When chatting with zephyr-7b-dpo-lora, as shown in the fig above , only the first 'Hello' was I sent, all the following content are generated by zephyr, including the \<user\> prompt. I cannot figure out why.

**Comment by DRXD1000** [comment]
Try loading the sft adapter first. Then merge the adapter into the base model and than load the dpo adapter.
U can use the following code:

model_name = "alignment-handbook/zephyr-7b-sft-lora"
tokenizer = AutoTokenizer.from_pretrained("alignment-handbook/zephyr-7b-sft-lora")
model = AutoPeftModelForCausalLM.from_pretrained(
    model_name,
    device_map="auto",
    use_flash_attention_2=True,
    torch_dtype = torch.bfloat16,
    use_cache=True
)
print("Merging Model")
model = model.merge_and_unload()
print("Model Merged")

peft_config = PeftConfig.from_pretrained("alignment-handbook/zephyr-7b-dpo-lora")

model_kwargs = dict(
    trust_remote_code=True,
    use_flash_attention_2=True,
    torch_dtype=torch.bfloat16,
    use_cache=True,
    device_map="auto"
)

# base_model = AutoModelForCausalLM.from_pretrained(
#     peft_config.base_model_name_or_path,
#     **model_kwargs,
# )
model = PeftModel.from_pretrained(model, "alignment-handbook/zephyr-7b-dpo-lora")

**Comment by DRXD1000** [comment]
Try loading the sft adapter first. Then merge the adapter into the base model and than load the dpo adapter.
U can use the following code:

model_name = "alignment-handbook/zephyr-7b-sft-lora"
tokenizer = AutoTokenizer.from_pretrained("alignment-handbook/zephyr-7b-sft-lora")
model = AutoPeftModelForCausalLM.from_pretrained(
    model_name,
    device_map="auto",
    use_flash_attention_2=True,
    torch_dtype = torch.bfloat16,
    use_cache=True
)
print("Merging Model")
model = model.merge_and_unload()
print("Model Merged")

peft_config = PeftConfig.from_pretrained("alignment-handbook/zephyr-7b-dpo-lora")

model = PeftModel.from_pretrained(model, "alignment-handbook/zephyr-7b-dpo-lora")

**Comment by 5thGenDev** [comment]
> Try loading the sft adapter first. Then merge the adapter into the base model and than load the dpo adapter. U can use the following code:

Where do I put these code in run_dpo.py?

**Comment by tommycwh** [comment]
I also got weird results from the provided ```alignment-handbook/zephyr-7b-dpo-qlora``` model. I tried to load the model using the method suggested above and another method that I usually used. I got different styles of generated texts but both are strange.

## Test data
I am testing with data from the ```test_prefs``` split of the ```HuggingFaceH4/ultrafeedback_binarized``` dataset, for example:
```
{'prompt': 'In this task, you are given a second sentence. Your task is to generate the first sentence on the same topic but incoherent and inconsistent with the second sentence.\n\nQ: Additionally , some groups may contain other specialists , such as a heavy weapons or language expert .\n\nA: Each squad member is specially trained as a weapons expert , medic , combat engineer or communications expert , respectively .\n****\nQ: However , the General Accounting Office identified 125 countries that received U.S. training and assistance for their police forces during fiscal year 1990 at a cost of at least $117 million .\n\nA: No government agency is in charge of calculating the cost .\n****\nQ: But his frozen body was found in the ice in Charlotte ( Rochester ) early the next spring by Silas Hudson .\n\nA:',
 'chosen': [{'content': 'In this task, you are given a second sentence. Your task is to generate the first sentence on the same topic but incoherent and inconsistent with the second sentence.\n\nQ: Additionally , some groups may contain other specialists , such as a heavy weapons or language expert .\n\nA: Each squad member is specially trained as a weapons expert , medic , combat engineer or communications expert , respectively .\n****\nQ: However , the General Accounting Office identified 125 countries that received U.S. training and assistance for their police forces during fiscal year 1990 at a cost of at least $117 million .\n\nA: No government agency is in charge of calculating the cost .\n****\nQ: But his frozen body was found in the ice in Charlotte ( Rochester ) early the next spring by Silas Hudson .\n\nA:',
   'role': 'user'},
  {'content': 'Could you provide some context or information about what you are looking for or any particular questions you have, so I can assist better?',
   'role': 'assistant'}],
...}
```

## DPO - Method 1
The model is prompted using the following code:
```
chosen_prompt_message = example["chosen"][0]
chosen_answer_message = example["chosen"][1]

chosen_prompt = tokenizer.apply_chat_template(
    [chosen_prompt_message],
    tokenize=False,
    add_generation_prompt=True,
)
inputs = tokenizer(chosen_prompt, return_tensors='pt').to(model.device)
generated_ids = model.generate(
    **inputs, eos_token_id=tokenizer.eos_token_id, max_new_tokens=4096,
    do_sample=False,
)
generated_texts = tokenizer.batch_decode(
    generated_ids[:, inputs['input_ids'].size(1) :],
    skip_special_tokens=True,
    clean_up_tokenization_spaces=True,
)[0]
```
Note: whether enabling or not ```do_sample``` does not seem to make a difference here. 

Using the method above to load the model, the model generate texts like:
```
'<|hudson|>\n<|silas|>\n<|charlotte|>\n<|rochester|>\n<|spring|>\n<|early|>\n<|next|>\n<|body|>\n<|frozen|>\n<|found|>\n<|ice|>\n<|1800|>\n<|1900|>\n<|1700|>\n<|1600|>\n<|1500|>\n...'
```
and the generation of numerical tokens does not stop until the max limit is hit.

## DPO - Method 2
The other method I used to load the model is:
```
import torch
from transformers import AutoModelForCausalLM, AutoTokenizer
from transformers import BitsAndBytesConfig

quantization_config = BitsAndBytesConfig(
    load_in_4bit=True,
    bnb_4bit_compute_dtype=torch.float16,
    bnb_4bit_quant_type="nf4",
    bnb_4bit_use_double_quant=True,
)

model = AutoModelForCausalLM.from_pretrained(
    model_id,
    device_map="auto",
    quantization_config=quantization_config,
)
tokenizer = AutoTokenizer.from_pretrained(
    model_id,
)
```
Note: loading in ```torch.bfloat16``` or ```torch.float16``` does not make a difference here.

The result looks something like:
```
In this task, you are given a second sentence. Your task is to generate the first sentence on the same topic but incoherent and inconsistent with the second sentence.

Q: The first step is to create a new class called MyClass.

A: The class MyClass is a subclass of the class Object.
****
Q: The first step is to create a new class called MyClass.

A: The class MyClass is a subclass of the class Object.
****
Q: The first step is to create a new class called MyClass.

A: The class MyClass is a subclass of the class Object.
****
...
```
Again, the generation does not stop till the max limit is hit.

## SFT
As a reference, I tried to do inference with the provided sft model ```alignment-handbook/zephyr-7b-sft-full``` and the result seems to be more reasonable. Using Method 2 to load the model, the model generate texts like:
```
The frozen body of the man was discovered by Silas Hudson in the ice in Charlotte ( Rochester ) early the next spring.
```
The model knows when to stop generating texts.

## Question
I see some other issues in this repository reporting that metric-wise the performance degrades after performing dpo by oneself. However, here I am using the provided dpo model and the results still seems strange. Is this the behavior expected from the provided dpo model, or is there anything that I am missing that causes this problem? 

Thank you.
